helix-core/src/main/java/org/apache/helix/task/TaskStateModelFactory.java

```diff
@@ -102,4 +107,40 @@ public boolean isShutdown() {
   public boolean isTerminated() {
     return _taskExecutor.isTerminated();
   }
+
+  /*
+   * Get target thread pool size from InstanceConfig first; if that fails, get it from
+   * ClusterConfig; if that fails, fall back to the default value.
+   */
+  private static int _getTaskThreadPoolSize(HelixManager manager) {
+    ConfigAccessor configAccessor = manager.getConfigAccessor();
+    // Check instance config first for thread pool size
+    try {
+      InstanceConfig instanceConfig =
+          configAccessor.getInstanceConfig(manager.getClusterName(), manager.getInstanceName());
+      int targetTaskThreadPoolSize = instanceConfig.getTargetTaskThreadPoolSize();
+      if (_verifyTargetThreadPoolSize(targetTaskThreadPoolSize)) {
+        return targetTaskThreadPoolSize;
+      }
+    } catch (HelixException e) {
+      // Pass if InstanceConfig doesn't exist.
+    }
+
+    // Fallback to cluster config since instance config doesn't provide the value
+    try {
+      ClusterConfig clusterConfig = configAccessor.getClusterConfig(manager.getClusterName());
+      int targetTaskThreadPoolSize = clusterConfig.getTargetTaskThreadPoolSize();
+      if (_verifyTargetThreadPoolSize(targetTaskThreadPoolSize)) {
+        return targetTaskThreadPoolSize;
+      }
+    } catch (HelixException e) {
+      // Pass if ClusterConfig doesn't exist.
+    }
+
+    return DEFAULT_TASK_THREAD_POOL_SIZE;
+  }
+
+  private static boolean _verifyTargetThreadPoolSize(int targetTaskThreadPoolSize) {
```

Comment:
You could do this in the setter?

Reply:
This is to check against the default value: when the field is not specified, a -1 is returned from `getTargetTaskThreadPoolSize()`. This check makes sure the default value is filtered out. 

I don't see any other good way to do `getIntField()` while handling the "not defined" case differently, since `getIntField()` always asks for a default value.

Reply:
@narendly Reminder on this conversation. 

I want to know if there's a good way to use `getIntField()` while handling the "not defined" case differently. If not, I will proceed with the current design.